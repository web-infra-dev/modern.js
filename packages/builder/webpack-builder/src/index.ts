export { createBuilder } from './core/createBuilder';
export { mergeBuilderConfig } from './shared/utils';

export type { BuilderInstance } from './core/createBuilder';
export type {
  // Plugin Types
  BuilderPlugin,
  BuilderContext,
  BuilderOptions,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CssExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
