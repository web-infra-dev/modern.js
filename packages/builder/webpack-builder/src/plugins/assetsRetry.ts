import { BuilderPlugin } from '../types/plugin';

export function PluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'webpack-builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getBuilderConfig();
        if (!config.output?.assetsRetry || target === 'node') {
          return;
        }
        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
