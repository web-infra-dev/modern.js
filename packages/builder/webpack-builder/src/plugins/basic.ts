import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const PluginBasic = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.set('javascript', {
        exportsPresence: 'error',
      });
    });
  },
});
