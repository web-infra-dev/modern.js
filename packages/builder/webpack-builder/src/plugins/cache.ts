import { join } from 'path';
import { isFileExists } from '../shared';
import type { BuilderPlugin } from '../types';

export const PluginCache = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-cache',

  setup(api) {
    api.modifyWebpackChain(async (chain, { target, env }) => {
      const { context } = api;
      const cacheDirectory = join(context.cachePath, 'webpack');
      const rootPackageJson = join(context.rootPath, 'package.json');
      const browserslistConfig = join(context.rootPath, '.browserslistrc');

      /**
       * webpack can't detect the changes of framework config, tsconfig and browserslist config.
       * but they will affect the compilation result, so they need to be added to buildDependencies.
       */
      const buildDependencies: Record<string, string[]> = {
        packageJson: [rootPackageJson],
      };
      if (context.configPath) {
        buildDependencies.config = [context.configPath];
      }
      if (context.tsconfigPath) {
        buildDependencies.tsconfig = [context.tsconfigPath];
      }
      if (await isFileExists(browserslistConfig)) {
        buildDependencies.browserslistrc = [browserslistConfig];
      }

      chain.cache({
        // The default cache name of webpack is '${name}-${env}', and the `name` is `default` by default.
        // We set cache name to avoid cache conflicts of different targets.
        name: `${target}-${env}`,
        type: 'filesystem',
        cacheDirectory,
        buildDependencies,
      });
    });
  },
});
