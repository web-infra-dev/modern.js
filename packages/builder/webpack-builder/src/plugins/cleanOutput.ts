import type { BuilderPlugin } from '../types';

export const PluginCleanOutput = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-clean-output',

  setup(api) {
    const config = api.getBuilderConfig();

    if (config.output?.cleanDistPath) {
      const clean = async () => {
        const { emptyDir } = await import('@modern-js/utils');
        const { distPath } = api.context;
        await emptyDir(distPath);
      };

      api.onBeforeBuild(clean);
      api.onBeforeStartDevServer(clean);
    }
  },
});
