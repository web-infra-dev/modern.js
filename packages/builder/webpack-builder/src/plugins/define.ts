import _ from '@modern-js/utils/lodash';
import type { BuilderPlugin, SourceConfig } from '../types';

export const PluginDefine = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-define',

  async setup(api) {
    const { default: webpack } = await import('webpack');
    const config = api.getBuilderConfig();

    const builtinVars: NonNullable<SourceConfig['globalVars']> = {
      NODE_ENV: process.env.NODE_ENV || 'development',
    };

    // Serialize global vars. User can customize value of `builtinVars`.
    const globalVars = _.assign(builtinVars, config.source?.globalVars);
    const serializedVars = _(globalVars)
      .mapKeys((_value, key) => `process.env.${key}`)
      .mapValues(value => JSON.stringify(value))
      .value();
    // Macro defines.
    const defineExprs = config.source?.define ?? {};

    // Apply define plugin
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      chain
        .plugin(CHAIN_ID.PLUGIN.DEFINE)
        .use(webpack.DefinePlugin, [{ ...serializedVars, ...defineExprs }]);
    });
  },
});
