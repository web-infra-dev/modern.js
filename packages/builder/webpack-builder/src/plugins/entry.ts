import type { BuilderPlugin } from '../types';

export const PluginEntry = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-entry',

  setup(api) {
    api.modifyWebpackChain(async chain => {
      const { ensureArray } = await import('@modern-js/utils');

      const { entry } = api.context;
      const config = api.getBuilderConfig();

      const { preEntry } = config.source || {};
      const preEntries = preEntry ? ensureArray(preEntry) : [];

      Object.keys(entry).forEach(entryName => {
        preEntries.forEach(entry => {
          chain.entry(entryName).add(entry);
        });

        ensureArray(entry[entryName]).forEach(item => {
          chain.entry(entryName).add(item as string);
        });
      });
    });
  },
});
