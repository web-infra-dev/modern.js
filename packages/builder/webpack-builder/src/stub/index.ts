export { createStubBuilder } from './builder';
export { createStubContext } from './context';
