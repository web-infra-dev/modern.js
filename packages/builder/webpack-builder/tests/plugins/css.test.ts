import { expect, describe, it } from 'vitest';
import { PluginCss } from '../../src/plugins/css';
import { PluginSass } from '../../src/plugins/sass';
import { PluginLess } from '../../src/plugins/less';
import { createStubBuilder } from '../../src/stub';

describe('plugins/css', () => {
  // skipped because this case time out in CI env
  it.skip('should set css config with style-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss()],
      builderConfig: {
        tools: {
          styleLoader: {},
        },
      },
    });
    const includeStyleLoader = await builder.matchWebpackLoader(
      'style-loader',
      'index.css',
    );

    expect(includeStyleLoader).toBe(true);
  });

  // skipped because this case time out in CI env
  it.skip('should set css config with mini-css-extract-plugin', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss()],
      builderConfig: {},
    });

    const includeMiniCssExtractLoader = await builder.matchWebpackLoader(
      'mini-css-extract-plugin',
      'index.css',
    );

    expect(includeMiniCssExtractLoader).toBe(true);
  });

  it('should add sass-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginSass()],
      builderConfig: {
        tools: {
          sass: {},
        },
      },
    });

    const includeSassLoader = await builder.matchWebpackLoader(
      'sass-loader',
      'index.scss',
    );

    expect(includeSassLoader).toBe(true);
  });

  it('should add less-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginLess()],
      builderConfig: {
        tools: {
          less: {},
        },
      },
    });

    const includeLessLoader = await builder.matchWebpackLoader(
      'less-loader',
      'index.less',
    );

    expect(includeLessLoader).toBe(true);
  });

  it('should override browserslist of autoprefixer when using output.overrideBrowserslist config', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCss()],
      builderConfig: {
        output: {
          overrideBrowserslist: ['Chrome 80'],
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
