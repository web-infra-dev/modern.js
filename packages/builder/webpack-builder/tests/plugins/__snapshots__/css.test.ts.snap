// Vitest Snapshot v1

exports[`plugins/css > should override browserslist of autoprefixer when using output.overrideBrowserslist config 1`] = `
{
  "module": {
    "rules": [
      {
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/packages/builder/webpack-builder/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "exportOnlyLocals": false,
                "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              },
              "sourceMap": false,
            },
          },
          {
            "loader": "<ROOT>/packages/builder/webpack-builder/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  [Function],
                  {
                    "browsers": [
                      "Chrome 80",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "Chrome 80",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                ],
              },
              "sourceMap": false,
            },
          },
        ],
      },
    ],
  },
}
`;
